proposals/exception-handling/Exceptions-formal-overview.md

```diff
@@ -0,0 +1,207 @@
+# 3rd proposal formal spec overview
+
+This is an overview of the 3rd proposal's formal spec additions, to aid in discussions concerning the proposed semantics.
+
+## Abstract Syntax
+
+### Types
+
+#### Tag Types
+
+```
+tagtype ::= [t*]→[]
+```
+
+### Instructions
+
+```
+instr ::= ... | throw tagidx | rethrow labelidx
+        | try blocktype instr* (catch tagidx instr*)* (catch_all instr*)? end
+        | try blocktype instr* delegate labelidx
+```
+
+### Modules
+
+#### Tags
+
+```
+tag ::= export* tag tagtype  | export* tag tagtype import
+```
+
+#### Modules
+
+```
+mod ::= module ... tag*
+```
+
+## Validation (Typing)
+
+#### Modification to labels
+
+To verify that the `rethrow l` instruction refers to a label surrounding the instructions of a catch block (call this a catch-label), we introduce a `kind` attribute to labels in the validation context, which is set to `catch` when the label is a catch-label and empty otherwise.
+
+```
+labelkind ::= catch
+labeltype ::= {result resulttype, kind labelkind?}
+C ::= {..., labels labeltype}
+```
+
+The original notation `labels [t*]` is now an abbreviation for:
+
+```
+labels [t*] ::= labels {result [t*], kind ε}
+```
+
+
+### Instructions
+
+
+```
+C.tags[x] = [t*]→[]
+-----------------------------
+C ⊢ throw x : [t1* t*]→[t2*]
+
+
+C.labels[l].kind = catch
+----------------------------
+C ⊢ rethrow l : [t1*]→[t2*]
+
+
+C.types[bt] = [t1*]→[t2*]
+C, labels [t2*] ⊢ instr* : [t1*]→[t2*]
+(C.tags[x] = [t*]→[] ∧
+ C, labels { result [t2*], kind catch } ⊢ instr'* : [t*]→[t2*])*
+(C, labels { result [t2*], kind catch } ⊢ instr''* : []→[t2*])?
+-----------------------------------------------------------------------------
+C ⊢ try bt instr* (catch x instr'*)* (catch_all instr''*)? end : [t1*]→[t2*]
+
+
+C.types[bt] = [t1*]→[t2*]
+C, labels [t2*] ⊢ instr* : [t1*]→[t2*]
+C.labels[l] = [t*]
+-------------------------------------------
+C ⊢ try bt instr* delegate l : [t1*]→[t2*]
+```
+
+## Execution (Reduction)
+
+### Runtime structure
+
+#### Stores
+
+```
+S ::= {..., tags taginst*}
+```
+
+#### Tag Instances
+
+```
+taginst ::= {type tagtype}
+```
+
+#### Module Instances
+
+```
+m ::= {..., tags tagaddr*}
+```
+
+#### Administrative Instructions
+
+```
+instr ::= ... | throw tagaddr | catch_n{ tagaddr? instr* }* instr* end
+        | delegate{ labelidx } instr* end | caught_m{ tagaddr val^n } instr* end
```

Comment:
These two instructions are quite different. In short, a `catch_all` reduces to a `catch_m` which only reduces to a `caught_m` if it contains a throw.

In a bit more detail, an instruction `try ... catch_all instr* end` reduces to

```
label_m{} ( catch_m{ ε instr* } ... end ) end
```

Only if the (throw context) `...` contains a `val^n (throw x)` can we reduce to this `caught`:

```
label_m{} ( caught_m{ a_x val^n} instr* end ) end
```

Now `caught` holds the exception that was caught by that `catch_all`, in case a `rethrow` needs to access the caught exception. Note that due to validation `rethrow` can only target `caught` instructions. Does that make sense? Please let me know if this explanation wasn't clear enough.

Reply:
I think I understand. I forgot about the possibility of rethrowing. Thanks!